micronota/workflow.py

```diff
@@ -0,0 +1,78 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2015--, micronota development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from os.path import splitext, basename, join
+from importlib import import_module
+
+from skbio.sequence import IntervalMetadata
+from skbio import read, DNA
+
+from .util import _tmp_file
+from . import bfillings
+
+
+def annotate(in_fp, in_fmt, out_dir, out_fmt, cpus, config):
+    '''Annotate the sequences in the input file.'''
+    fn = splitext(basename(in_fp))[0]
+    # store annotated seq file.
+    out_fp = join(out_dir, '%s.gbk' % fn)
+    # append the file for multiple sequences
+    out = open(out_fp, 'a')
+    sec = 'WORKFLOW'
+    workflows = config[sec]['order'].split(' > ')
+    for seq in read(in_fp, format=in_fmt):
+        # interval_metadata
+        im = {}
+        with _tmp_file() as f:
+            _, fp = f
+            seq.write(fp, format='fasta')
+            # output dir for the current input seq
+            out_dir_seq = join(out_dir, seq.metadata['id'])
+            for k in workflows:
+                tasks = config[sec][k].split(' > ')
+                if k == 'features':
+                    im_ = identify_all_features(fp, out_dir_seq, tasks, config)
+                    im.update(im_)
+                if k == 'CDS':
+                    annotate_all_cds()
+        seq.interval_metadata = IntervalMetadata(im)
+        seq.write(out, format=out_fmt)
+    out.close()
+
+
+def identify_all_features(fp, out_dir, tasks, config,
+                          id_func='identify_features',
+                          parse_func='parse_output'):
+    '''Identify all the features on the sequence in the input file.
+
+    It runs thru all the tasks specified in sequential order.
+
+    Parameters
+    ----------
+
+    Returns
+    -------
+    '''
+    im = {}
+    for task in tasks:
+        items = task.split(':')
+        tool = items[0]
+        submodule = import_module('.%s' % tool, bfillings.__name__)
```

Comment:
just reminded me of seeing log messages about loading and unloading modules
i guess

On Wed, Feb 24, 2016 at 10:30 PM, Zech Xu <notifications@github.com> wrote:

> In micronota/workflow.py
> <https://github.com/biocore/micronota/pull/46#discussion_r54053879>:
>
> > +                          parse_func='parse_output'):
> > +    '''Identify all the features on the sequence in the input file.
> > +
> > +    It runs thru all the tasks specified in sequential order.
> > +
> > +    Parameters
> > +    ----------
> > +
> > +    Returns
> > +    -------
> > +    '''
> > +    im = {}
> > +    for task in tasks:
> > +        items = task.split(':')
> > +        tool = items[0]
> > +        submodule = import_module('.%s' % tool, bfillings.__name__)
>
> yeah, plan to use logging for all the messages. curious why you brought
> it up at this line of code?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/micronota/pull/46/files#r54053879>.
>